src/Routing/Parser.purs

```diff
@@ -1,46 +1,45 @@
-module Routing.Parser (
-  parse
-  ) where
+module Routing.Parser (parse) where
 
-import Control.MonadPlus (guard)
+import Prelude
+
+import Routing.Types (Route, RoutePart(..))
 import Data.Array as A
-import Data.Either (fromRight)
-import Data.List (fromFoldable, List)
 import Data.Map as M
-import Data.Maybe (Maybe, fromMaybe)
 import Data.String as S
-import Data.String.Regex (Regex, regex, split) as R
-import Data.String.Regex.Flags (noFlags) as R
+import Control.MonadPlus (guard)
+import Data.List as L
+import Data.Maybe (Maybe(..))
 import Data.Traversable (traverse)
 import Data.Tuple (Tuple(..))
-import Partial.Unsafe (unsafePartial)
-import Prelude (map, discard, (>>>), ($), (<<<), (==), (<*>), (<$>), (<=))
-import Routing.Types (Route, RoutePart(..))
 
--- | Parse part of hash. Will return `Query (Map String String)` for query
+-- | Parse query part of hash. Will return `Map String String` for query
 -- | i.e. `"?foo=bar&bar=baz"` -->
--- |     `Query (fromList [Tuple "foo" "bar", Tuple "bar" "baz"])`
-parsePart :: String -> RoutePart
-parsePart str = fromMaybe (Path str) do
-  guard $ S.take 1 str == "?"
-  map (Query <<< M.fromFoldable)
-    $ traverse part2tuple parts
+-- |     `fromList [Tuple "foo" "bar", Tuple "bar" "baz"]`
+parseQueryPart :: (String -> String) -> String -> Maybe (M.Map String String)
+parseQueryPart decoder =
+  map M.fromFoldable <<< traverse part2tuple <<< S.split (S.Pattern "&")
   where
-  parts :: List String
-  parts = fromFoldable $ S.split (S.Pattern "&") $ S.drop 1 str
-
-  part2tuple :: String -> Maybe (Tuple String String)
-  part2tuple input = do
-    let keyVal = S.split (S.Pattern "=") input
-    guard $ A.length keyVal <= 2
-    Tuple <$> (A.head keyVal) <*> (keyVal A.!! 1)
-
-
-splitRegex :: R.Regex
-splitRegex = unsafePartial fromRight $ R.regex "\\/|(?=\\?)" R.noFlags
+    part2tuple :: String -> Maybe (Tuple String String)
+    part2tuple input = do
+      let keyVal = decoder <$> S.split (S.Pattern "=") input
+      guard $ A.length keyVal <= 2
+      Tuple <$> A.head keyVal <*> keyVal A.!! 1
```

Comment:
Do query parts _necessarily_ need to contain two parts?

Do we want to be able to handle? (Might totally not be worth it because of gnarly types):
`"?foo&bar=baz"`

Reply:
I think initial motivation of omitting `?foo&bar=baz` was `/foo/?bar=baz` 
But yeah handling `?foo&bar` probably has sense, maybe query parser from uri could help.